File: src/main/kotlin/demo/kotliln/boardgame/service/rating/RatingEndpoint.java
package demo.kotliln.boardgame.service.rating;

import demo.kotliln.boardgame.api.models.BoardGame;
import demo.kotliln.boardgame.api.requests.RatingRequest;
import demo.kotliln.boardgame.service.exceptions.BoardGameNotFound;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.List;

import static org.springframework.web.bind.annotation.RequestMethod.GET;
import static org.springframework.web.bind.annotation.RequestMethod.POST;

@Slf4j
@RestController
public class RatingEndpoint {

    private final BoardGameRatingService ratingService;

    @Autowired
    public RatingEndpoint(final BoardGameRatingService ratingService) {
        this.ratingService = ratingService;
    }

    @RequestMapping(
            path = "/rate",
            method = POST,
            consumes = "application/json"
    )
    public ResponseEntity<BoardGame> rate(@RequestBody final RatingRequest ratingRequest) {

        log.info("Rating request coming in for game {} with rate {}",
                ratingRequest.getRatedGame(),
                ratingRequest.getRate()
        );

        try {
            final RateBoardGame rateBoardGame = new RateBoardGame(
                    ratingRequest.getRatedGame(),
                    ratingRequest.getRate()
            );

            final BoardGame ratedBoardGame = ratingService.ratingBoardGame(rateBoardGame);

            return ResponseEntity.ok().body(ratedBoardGame);
        } catch (BoardGameNotFound boardGameNotFound) {
            return ResponseEntity.notFound().build();
        } catch (RuntimeException re) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build();
        }

    }

    @RequestMapping(
            path = "/boardgames",
            method = GET,
            consumes = "application/json"
    )
    public ResponseEntity<List<BoardGame>> boardgamesForMinimalMeanRate(
            @RequestParam(name = "rate") final Double rate
    ) {
        List<BoardGame> boardGames = ratingService.withHigherRateThan(rate);
        return ResponseEntity.ok().body(boardGames);
    }
}

File: src/main/kotlin/demo/kotliln/boardgame/service/rating/controllers/RatingEndpoint.kt
package demo.kotliln.boardgame.service.rating.controllers

import demo.kotliln.boardgame.api.models.BoardGame
import demo.kotliln.boardgame.api.requests.RatingRequest
import demo.kotliln.boardgame.service.boardgame.exceptions.BoardGameNotFound
import demo.kotliln.boardgame.service.rating.BoardGameRatingService
import demo.kotliln.boardgame.service.rating.entities.RateBoardGame
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*

@RestController
class RatingEndpoint (private val boardGameRatingService: BoardGameRatingService) {

    @RequestMapping(path = ["/rate"], method = [RequestMethod.POST], consumes = ["application/json"])
    fun rate(@RequestBody ratingRequest: RatingRequest): ResponseEntity<BoardGame> {
        try {
            val rateBoardGame = RateBoardGame(ratingRequest.ratedGame, ratingRequest.rate)
            val ratedBoardGame = boardGameRatingService.ratingBoardGame(rateBoardGame)
            return ResponseEntity.ok().body(ratedBoardGame)
        } catch (boardGameNotFound: BoardGameNotFound) {
            return ResponseEntity.notFound().build()
        } catch (re: RuntimeException) {
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).build()
        }
    }

    @RequestMapping(path = ["/boardgames"], method = [RequestMethod.GET], consumes = ["application/json"])
    fun boardgamesForMinimalMeanRate(@RequestParam(name = "rate") rate: Double): ResponseEntity<List<BoardGame>> {
        val boardGames = boardGameRatingService.withHigherRateThan(rate)
        return ResponseEntity.ok().body(boardGames)
    }

}